packages/firestore/src/model/proto_values.ts

```diff
@@ -0,0 +1,385 @@
+    case TypeOrder.ObjectValue:
+      return compareMaps(left.mapValue!, right.mapValue!);
+    default:
+      throw fail('Invalid value type: ' + leftType);
+  }
+}
+
+function compareBooleans(b1: boolean, b2: boolean): number {
+  return b1 !== b2 ? (b1 ? 1 : -1) : 0;
+}
+
+function compareNumbers(left: api.Value, right: api.Value): number {
+  const leftNumber =
+    'doubleValue' in left
+      ? normalizeNumber(left.doubleValue)
+      : normalizeNumber(left.integerValue);
+  const rightNumber =
+    'doubleValue' in right
+      ? normalizeNumber(right.doubleValue)
+      : normalizeNumber(right.integerValue);
+  return numericComparator(leftNumber, rightNumber);
+}
+
+function compareTimestamps(
+  left: string | { seconds?: string; nanos?: number },
+  right: string | { seconds?: string; nanos?: number }
+): number {
+  const leftTimestamp = normalizeTimestamp(left);
+  const rightTimestamp = normalizeTimestamp(right);
+
+  const comparison = primitiveComparator(
+    leftTimestamp.seconds,
+    rightTimestamp.seconds
+  );
+  if (comparison !== 0) {
+    return comparison;
+  }
+  return primitiveComparator(leftTimestamp.nanos, rightTimestamp.nanos);
+}
+
+function compareReferences(leftPath: string, rightPath: string): number {
+  const leftSegments = leftPath.split('/');
+  const rightSegments = rightPath.split('/');
+  for (let i = 0; i < leftSegments.length && i < rightSegments.length; i++) {
+    const comparison = primitiveComparator(leftSegments[i], rightSegments[i]);
+    if (comparison !== 0) {
+      return comparison;
+    }
+  }
+  return primitiveComparator(leftSegments.length, rightSegments.length);
+}
+
+function compareGeoPoints(left: api.LatLng, right: api.LatLng): number {
+  const comparison = primitiveComparator(
+    normalizeNumber(left.latitude),
+    normalizeNumber(right.latitude)
+  );
+  if (comparison !== 0) {
+    return comparison;
+  }
+  return primitiveComparator(
+    normalizeNumber(left.longitude),
+    normalizeNumber(right.longitude)
+  );
+}
+
+function compareBlobs(
+  left: string | Uint8Array,
+  right: string | Uint8Array
+): number {
+  const leftBytes = normalizeByteString(left);
+  const rightBytes = normalizeByteString(right);
+  return leftBytes.compareTo(rightBytes);
+}
+
+function compareArrays(left: api.ArrayValue, right: api.ArrayValue): number {
+  const leftArray = left.values || [];
+  const rightArray = right.values || [];
+
+  for (let i = 0; i < leftArray.length && i < rightArray.length; ++i) {
+    const valueCompare = compare(leftArray[i], rightArray[i]);
+    if (valueCompare) {
+      return valueCompare;
+    }
+  }
+  return primitiveComparator(leftArray.length, rightArray.length);
+}
+
+function compareMaps(left: api.MapValue, right: api.MapValue): number {
+  const leftMap = left.fields || {};
+  const leftKeys = keys(leftMap);
+  const rightMap = right.fields || {};
+  const rightKeys = keys(leftMap);
+
+  leftKeys.sort();
```

Comment:
Are these sorts required? The protos we get from the backend are guaranteed to come with keys in order, and JS guarantees that traversing keys happens in the order they're added to the object.